File: src/connectors/otx.js
import { BaseConnector } from './base.js';
import https from 'https';

export class OTXConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'AlienVault OTX';
        this.apiKey = process.env.OTX_API_KEY;
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];

            if (inputType === 'domain') {
                const general = await this._apiGet(`/api/v1/indicators/domain/${inputValue}/general`);
                if (general) {
                    if (general.pulse_info?.count > 0) {
                        entities.push({
                            type: 'domain',
                            data: {
                                name: inputValue,
                                pulse_count: general.pulse_info.count,
                                pulses: general.pulse_info.pulses?.slice(0, 5).map(p => ({ name: p.name, created: p.created })),
                                source: 'otx'
                            },
                            confidence: 0.85
                        });
                    }
                }

                const passive = await this._apiGet(`/api/v1/indicators/domain/${inputValue}/passive_dns`);
                if (passive?.passive_dns) {
                    const seenIPs = new Set();
                    for (const rec of passive.passive_dns.slice(0, 30)) {
                        if (rec.address && !seenIPs.has(rec.address)) {
                            seenIPs.add(rec.address);
                            entities.push({
                                type: 'ip',
                                data: {
                                    address: rec.address,
                                    first_seen: rec.first,
                                    last_seen: rec.last,
                                    record_type: rec.record_type,
                                    source: 'otx_passive_dns'
                                },
                                confidence: 0.7,
                                asKnown: { type: 'ip', value: rec.address }
                            });
                        }
                    }
                }

                const malware = await this._apiGet(`/api/v1/indicators/domain/${inputValue}/malware`);
                if (malware?.data?.length) {
                    entities.push({
                        type: 'domain',
                        data: {
                            name: inputValue,
                            malware_samples: malware.data.length,
                            recent_hashes: malware.data.slice(0, 5).map(m => m.hash),
                            source: 'otx_malware'
                        },
                        confidence: 0.9
                    });
                }
            }

            if (inputType === 'ip') {
                const general = await this._apiGet(`/api/v1/indicators/IPv4/${inputValue}/general`);
                if (general) {
                    entities.push({
                        type: 'ip',
                        data: {
                            address: inputValue,
                            asn: general.asn,
                            country: general.country_name,
                            pulse_count: general.pulse_info?.count || 0,
                            reputation: general.reputation,
                            source: 'otx'
                        },
                        confidence: 0.85
                    });
                }

                const passive = await this._apiGet(`/api/v1/indicators/IPv4/${inputValue}/passive_dns`);
                if (passive?.passive_dns) {
                    const seenDomains = new Set();
                    for (const rec of passive.passive_dns.slice(0, 30)) {
                        if (rec.hostname && !seenDomains.has(rec.hostname)) {
                            seenDomains.add(rec.hostname);
                            entities.push({
                                type: 'domain',
                                data: {
                                    name: rec.hostname,
                                    first_seen: rec.first,
                                    last_seen: rec.last,
                                    source: 'otx_passive_dns'
                                },
                                confidence: 0.7,
                                asKnown: { type: 'domain', value: rec.hostname }
                            });
                        }
                    }
                }
            }

            if (inputType === 'url') {
                const general = await this._apiGet(`/api/v1/indicators/url/${encodeURIComponent(inputValue)}/general`);
                if (general) {
                    entities.push({
                        type: 'url',
                        data: {
                            url: inputValue,
                            pulse_count: general.pulse_info?.count || 0,
                            source: 'otx'
                        },
                        confidence: 0.8
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _apiGet(path) {
        return new Promise((resolve, reject) => {
            const headers = { 'Accept': 'application/json' };
            if (this.apiKey) headers['X-OTX-API-KEY'] = this.apiKey;

            const req = https.get({
                hostname: 'otx.alienvault.com',
                path,
                headers,
                timeout: 15000
            }, res => {
                if (res.statusCode === 404) { res.resume(); return resolve(null); }
                if (res.statusCode === 429) { res.resume(); return reject(new Error('OTX rate limited')); }
                if (res.statusCode >= 400) { res.resume(); return reject(new Error(`OTX HTTP ${res.statusCode}`)); }
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => {
                    try { resolve(JSON.parse(data)); } catch { resolve(null); }
                });
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('OTX timeout')); });
        });
    }

    async healthCheck() {
        return { ok: true, version: 'v1', hasApiKey: !!this.apiKey };
    }
}

File: cyclops/connectors/otx.py
import os
import aiohttp
from cyclops.connectors.base import BaseConnector


class OTXConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "AlienVault OTX"
        self.api_key = os.environ.get("OTX_API_KEY")

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            entities = []
            headers = {"Accept": "application/json"}
            if self.api_key:
                headers["X-OTX-API-KEY"] = self.api_key
            async with aiohttp.ClientSession(headers=headers) as session:
                if input_type == "domain":
                    passive = await self._get(session, f"/api/v1/indicators/domain/{input_value}/passive_dns")
                    seen = set()
                    for rec in (passive or {}).get("passive_dns", [])[:30]:
                        addr = rec.get("address")
                        if addr and addr not in seen:
                            seen.add(addr)
                            entities.append({"type": "ip", "data": {"address": addr, "first_seen": rec.get("first"), "last_seen": rec.get("last"), "source": "otx_passive_dns"}, "confidence": 0.7, "as_known": {"type": "ip", "value": addr}})
                    general = await self._get(session, f"/api/v1/indicators/domain/{input_value}/general")
                    if general and general.get("pulse_info", {}).get("count", 0) > 0:
                        entities.append({"type": "domain", "data": {"name": input_value, "pulse_count": general["pulse_info"]["count"], "source": "otx"}, "confidence": 0.85})
                elif input_type == "ip":
                    general = await self._get(session, f"/api/v1/indicators/IPv4/{input_value}/general")
                    if general:
                        entities.append({"type": "ip", "data": {"address": input_value, "asn": general.get("asn"), "country": general.get("country_name"), "pulse_count": general.get("pulse_info", {}).get("count", 0), "source": "otx"}, "confidence": 0.85})
                    passive = await self._get(session, f"/api/v1/indicators/IPv4/{input_value}/passive_dns")
                    seen = set()
                    for rec in (passive or {}).get("passive_dns", [])[:30]:
                        hostname = rec.get("hostname")
                        if hostname and hostname not in seen:
                            seen.add(hostname)
                            entities.append({"type": "domain", "data": {"name": hostname, "source": "otx_passive_dns"}, "confidence": 0.7, "as_known": {"type": "domain", "value": hostname}})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def _get(self, session, path):
        async with session.get(f"https://otx.alienvault.com{path}", timeout=aiohttp.ClientTimeout(total=15)) as resp:
            if resp.status in (404, 429) or resp.status >= 400:
                return None
            return await resp.json()

    async def health_check(self):
        return {"ok": True, "hasApiKey": bool(self.api_key)}
